### PR TITLE
Add configurable optimizer/schedluer

### PR DESCRIPTION
**Pull request by naomi-simumba**
Replaces the hardcoded AdamW + ReduceLROnPlateau in configure_optimizers() with config-driven selection. The optimizer type, its hyperparameters, the scheduler type, and its hyperparameters are all specified in the YAML optimizer block. AdamW remains the default when type is omitted; omitting scheduler_type disables the scheduler. The implementation is intentionally minimal so that existing configurations continue to behave exactly as before, while advanced users gain the ability to select alternative optimizers and schedulers through configuration rather than code changes.
addresses issue: https://github.com/gridfm/gridfm-graphkit/issues/24

`Rationale`: gridfm-graphkit is intended for use across different power grids, problem sizes, and research settings. Hardcoding the optimizer and scheduler means all users are constrained to the same training configuration regardless of their requirements. This PR removes that constraint through a minimal, backwards-compatible change.

- `Optimizer flexibility`: AdamW remains a strong default, but different workloads may require different optimisation strategies. Allowing optimizer selection through configuration avoids the need for code changes or downstream forks.
- `Scheduler flexibility`: ReduceLROnPlateau works well in many settings, but users may prefer alternatives such as cosine annealing, linear warmup, or no scheduler. The current implementation makes those choices unavailable without forking library code.
- `Alignment with common ML frameworks`: Frameworks such as PyTorch Lightning, Hugging Face Transformers, and Fairseq expose optimizers and schedulers through configuration. Following a similar pattern makes gridfm-graphkit easier to understand and integrate into existing ML workflows.
- `Reproducibility`: When optimizer and scheduler settings are defined in configuration, they become part of the experiment configuration captured by MLflow, making runs easier to reproduce and compare.

Example configs have been updated to an equivalent to the previous hardcoded behaviour
```
optimizer:
  type: AdamW
  learning_rate: 0.0005
  optimizer_params:
    betas: [0.9, 0.999]
    weight_decay: 0.01
  scheduler_type: ReduceLROnPlateau
  scheduler_params:
    mode: min
    factor: 0.7      # previously lr_decay
    patience: 5      # previously lr_patience
```

### CODE REVIEW

**Comment by naomi-simumba** [comment]
@romeokienzler  Would you mind taking a look at this when you have a chance? I'd be interested in your perspective on whether this is a useful extensibility point for downstream users of the library. Thanks

**Review by romeokienzler — COMMENTED** [review]
Thanks for this, @naomi-simumba — this is a genuinely useful change and the `getattr(torch.optim, …)` mapping approach is clean. Nicely done migrating all the example/test configs too, and dropping the `reduce_on_plateau`/`strict` flags is correct since Lightning auto-detects `ReduceLROnPlateau` and `strict` defaults to `True`. A few points from my review:

**Please address before merge (1):**
1. The "omit to get the default" behaviour described in the PR doesn't actually hold — it raises `AttributeError`. `NestedNamespace` extends `argparse.Namespace`, which raises on a missing key (no `__getattr__` returning `None`). So:
   - `if self.args.optimizer.type is None:` throws if `type` is *omitted* (rather than present-but-null) — it can never fall through to the AdamW default.
   - `self.args.optimizer.optimizer_params` and `.scheduler_params` are accessed directly, so omitting them throws too.
   - Interestingly `scheduler_type` **is** handled correctly via `getattr(self.args.optimizer, "scheduler_type", None)` — so the code is internally inconsistent. Could you use that same `getattr(..., default)` pattern for `type`, `optimizer_params`, and `scheduler_params`?

   This also affects the backward-compat claim: an existing config on the old schema (`beta1`/`beta2`/`lr_decay`/`lr_patience`) will break since none of the new keys exist. Migrating the in-repo configs hides this in CI but not for downstream users — worth either handling the old keys or flagging it as a breaking change in the description/changelog.

**Nice-to-have (2–3), not blocking:**
2. No test currently covers `configure_optimizers`. A small unit test (default path, custom optimizer, no-scheduler path, and the omitted-key path from #1) would lock in the intended contract.
3. Minor: the code reassigns `self.args.optimizer.type` / `.optimizer_params` on the live config (leaks into MLflow-logged params), and an invalid `type` string surfaces as a bare `AttributeError` on `torch.optim` rather than a clear "unknown optimizer" message.

Happy to approve once (1) is sorted. Thanks again!

**Comment by naomi-simumba** [comment]
Thank you for the thorough review 
I've addressed point #1: type, `optimizer_params`, and `scheduler_params` are now all handled with `getattr(..., default)`, consistent with how `scheduler_type` was already handled. Omitting any of them will now fall through to the correct default rather than raising. I also took the opportunity to add a clear `ValueError` for invalid optimizer and scheduler type strings , so users get a helpful message.
For point #2, I've added `tests/test_configure_optimizers.py` covering the default path, a custom optimizer, the no-scheduler path, the omitted-key paths, and the invalid type error cases.
Let me know if there's anything else you'd like me to adjust!

**Review by romeokienzler — APPROVED** [review]
Thanks @naomi-simumba — the updated version fully addresses the blocker: `type`, `optimizer_params`, and `scheduler_params` now use `getattr(..., default)` consistently, so omitting them no longer raises. You also nailed the nice-to-haves — clear `ValueError`s for unknown optimizer/scheduler types, no more in-place config mutation, and a focused 9-case test suite (`tests/test_configure_optimizers.py`) covering the default/omitted-key paths and the error messages. pytests are green. Approving.

Note: the only remaining red check is `pip-audit (deps)` on `lightning` (PYSEC-2026-3624) — that's a pre-existing, repo-wide issue unrelated to this PR, tracked in #103.

**Comment by romeokienzler** [comment]
Post-merge integration test on Vela (single H100, torch 2.12.1) — main @ 0.8.1 vs. the frozen 0.8.0 calibration baseline. Both integration tests passed; every metric landed inside the 0.8.0 bounds on the first attempt, confirming the optimizer/scheduler refactor introduced **no numerical regression**.

**PF (`test_train_pf`)**
| Metric | Value | 0.8.0 bounds |
|---|---|---|
| PBE Mean | 0.41404 | [0.40990, 0.41818] ✅ |

**OPF (`test_train_opf`)**
| Metric | Value | 0.8.0 bounds |
|---|---|---|
| Mean optimality gap (%) | 1.5787 | [1.5629, 1.5945] ✅ |
| RMSE PG generators (MW) | 2.6503 | [2.6238, 2.6768] ✅ |
| Avg. active res. (MW) | 0.29152 | [0.28860, 0.29443] ✅ |
| Avg. reactive res. (MVar) | 0.10529 | [0.10424, 0.10634] ✅ |
| Mean Qg violation | 0.10337 | [0.10234, 0.10441] ✅ |
| Branch angle / thermal violations | 0.0 | [0.0, 0.0] ✅ |

pytest: `2 passed in 2117.91s`. Fingerprint mismatch warning fired as designed (0.8.1 vs 0.8.0 baseline) and proceeded.
